interp/interp_test.go

```diff
@@ -191,3 +190,341 @@ func TestPiecewiseConstantPredict(t *testing.T) {
 	leftYs := []float64{-0.5, 1.5, 1.5, 1.5, 1, 1}
 	testInterpolatorPredict(t, pc, testXs, leftYs, 0)
 }
+
+func TestPiecewiseCubic(t *testing.T) {
+	t.Parallel()
+	const (
+		h        = 1e-8
+		valueTol = 1e-13
+		derivTol = 1e-6
+	)
+	for i, test := range []struct {
+		xs []float64
+		f  func(float64) float64
+		df func(float64) float64
+	}{
+		{
+			xs: []float64{-1.001, 0.2, 2},
+			f:  func(x float64) float64 { return x * x },
+			df: func(x float64) float64 { return 2 * x },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 2},
+			f:  func(x float64) float64 { return x * x },
+			df: func(x float64) float64 { return 2 * x },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 10},
+			f:  func(x float64) float64 { return 1.5*x - 1 },
+			df: func(x float64) float64 { return 1.5 },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 10},
+			f:  func(x float64) float64 { return -1 },
+			df: func(x float64) float64 { return 0 },
+		},
+		{
+			xs: []float64{-1.1, 0.2, 0.99, 2.5, 2.99},
+			f:  math.Cos,
+			df: math.Sin,
```

Comment:
`cos'(x) = -sin(x)` and why does the test passes?

Reply:
Hmm. This is very strange. I'll check.

Reply:
The test passes because the interpolation algorithm doesn't care what function I got the (x, y, dy/dx) points from - it just interpolates those points.